BUG: Fix behavior of replace_list with mixed types.

hasan-yaman opened the pull request:
- [x] closes #40371 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] whatsnew entry

Commit:
BUG: Fix behavior of replace_list with mixed types.

mzeitlin11 commented on pandas/tests/series/methods/test_replace.py:
```diff
@@ -460,3 +460,17 @@ def test_str_replace_regex_default_raises_warning(self, pattern):
+    @pytest.mark.parametrize(
+        "s, to_replace, value, expected",
+        [
+            (pd.Series([1]), np.array([1.0]), [0], pd.Series([0])),
```
I think you can deduplicate these by using the `frame_or_series` fixture in `pandas/conftest.py`. Then just one test can be added in tests/frame (which is where we try to put tests using the `frame_or_series` fixture).

mzeitlin11 reviewed (commented):
Thanks for the pr, @hasan-yaman!

mzeitlin11 commented:
Looks like this caused some test errors elsewhere: https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=56659&view=logs&j=b516280a-8c75-541b-8ba0-ae13561718a7&t=a9217cb1-24f6-5625-717d-72085a70e1d3&l=41

dsaxton commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1659,3 +1659,17 @@ def test_replace_bytes(self, frame_or_series):
+    @pytest.mark.parametrize(
+        "s, to_replace, value, expected",
+        [
+            (DataFrame([1]), np.array([1.0]), [0], DataFrame([0])),
```
May want to test this for different container types for `to_replace` rather than only `np.array`

hasan-yaman replied:
I added lists too. Should I add more types?

Commit:
BUG: Revert changes in cast.py and refactor tests

hasan-yaman commented:
Thanks for the comments.

- I reverted changes in cast.py file since it turns out making changes in cast.py wasn't good idea. 
- I updated the tests.
- I need some help for the PR. 
https://github.com/pandas-dev/pandas/blob/9c4c82112f894d631b09873a64cc74d212b2041d/pandas/core/internals/blocks.py#L801-L804

Removing if statement here will solve the problem but i am wondering is there are better way to solve it?

jbrockmendel commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1659,3 +1658,25 @@ def test_replace_bytes(self, frame_or_series):
         tm.assert_equal(obj, expected)
+
+    @pytest.mark.parametrize(
+        "s, to_replace, value, expected",
```
can you give this a multi-letter name instead of 's'

simonjayhawkins commented:
Thanks @hasan-yaman for the PR. can you fix-up failing tests... it appears actual fix has been lost in a subsequent commit and add a release note (will be 1.2.5 but can add to doc\source\whatsnew\v1.2.4.rst in the first instance and move once doc\source\whatsnew\v1.2.5.rst is available)

jreback commented:
> Thanks @hasan-yaman for the PR. can you fix-up failing tests... it appears actual fix has been lost in a subsequent commit and add a release note (will be 1.2.5 but can add to doc\source\whatsnew\v1.2.4.rst in the first instance and move once doc\source\whatsnew\v1.2.5.rst is available)

this is just a tests, no whatsnew needed.

simonjayhawkins commented:
> this is just a tests, no whatsnew needed.

IIUC this PR a regression fix for #40371 but the fix, added in the first commit, has been lost.

hasan-yaman commented:
Actually, first commit was causing too many problems with the tests, I couldn't fix them. So, I reverted the changes from the first commit. 
I am open to any idea to fix the problem. (Maybe others who know pandas better than me can attempt the fix the problem with separate PR.)

simonjayhawkins commented:
> I am open to any idea to fix the problem. (Maybe others who know pandas better than me can attempt the fix the problem with separate PR.)

The regression was caused by #38097, so the fix probably involves code changed in that PR. cc @jbrockmendel

jreback commented:
> > this is just a tests, no whatsnew needed.
> 
> IIUC this PR a regression fix for #40371 but the fix, added in the first commit, has been lost.

what do you mean? if this doesn't have a patch then the tests shouldn't pass (or is this *already) fixed?

simonjayhawkins commented:
> > > this is just a tests, no whatsnew needed.
> > 
> > 
> > IIUC this PR a regression fix for #40371 but the fix, added in the first commit, has been lost.
> 
> what do you mean? if this doesn't have a patch then the tests shouldn't pass (or is this *already) fixed?

The tests are not passing.

simonjayhawkins commented:
to_replace is dispatched to _replace_list if is_list_liketo_replace)

_replace_list only accepts lists, but also need to make sure that the list is python floats

```
>>> import numpy as np
(list(np.array([1.0]))[0])

type(np.array([1.0]).tolist()[0])>>> 
>>> type(np.array([1.0])[0])
<class 'numpy.float64'>
>>> 
>>> type(list(np.array([1.0]))[0])
<class 'numpy.float64'>
>>> 
>>> type(np.array([1.0]).tolist()[0])
<class 'float'>
>>> 
```

or change _can_hold_element

simonjayhawkins commented:
> or change _can_hold_element

this maybe a better option since this is where we have the inconsistency

```
>>> 
>>> pd.core.dtypes.cast.can_hold_element(arr=pd.Series([1]).values, element=1.0)
True
>>> 
>>> pd.core.dtypes.cast.can_hold_element(
...     arr=pd.Series([1]).values, element=np.array([1.0])[0]
... )
False
>>> 
```

simonjayhawkins commented:
> > or change _can_hold_element
> 
> this maybe a better option since this is where we have the inconsistency

changed my mind again! since we have a mature backport branch and may not have any more patch releases on 1.2.x let's not make this change and just patch the code that was added in #38097 to convert a numpy array to a list of python objects.

simonjayhawkins committed:
Merge remote-tracking branch 'upstream/master' into replace-floats-with-integers

simonjayhawkins committed:
add patch, release note and rename test variables

simonjayhawkins committed:
parametrize

lbittarello commented on pandas/tests/frame/methods/test_replace.py:
```diff
@@ -1428,6 +1428,25 @@ def test_replace_bytes(self, frame_or_series):
+            ([1.0], [1], [0], [0.0]),
+        ],
+    )
+    @pytest.mark.parametrize("box", [list, tuple, np.array])
```
Would it be possible to test `pd.array` too? Better safe than sorry. :)

simonjayhawkins replied:
That would fail. EA is not a documented allowed type for .replace. to be fair, nor is numpy array or tuple. The safest way to restore the old undocumented behavior is to revert #38097, but not sure that a performance hit for users passing the correct type is the way forward.

Any container, even a list, with numpy scalars will still fail. This is a fix for a specific case for an arguably incorrect usage without a perf hit for correct usage.

simonjayhawkins commented:
@jreback can you take a look. not the best fix, but will be fixed properly ( I think) in #41644, which would be updated to remove the code added here on master straight after this is backported.

simonjayhawkins commented on pandas/core/internals/blocks.py:
```diff
@@ -779,6 +781,14 @@ def _replace_list(
+        # until the issue is fixed properly in can_hold_element
+
+        # error: "Iterable[Any]" has no attribute "tolist"
+        if hasattr(src_list, "tolist"):
```
I could maybe change this to a simple `isinstance(src_list, np.array)`

simonjayhawkins commented:
will merge (and backport) this shortly if no further comments

simonjayhawkins committed:
Merge remote-tracking branch 'upstream/master' into replace-floats-with-integers

simonjayhawkins committed:
Merge remote-tracking branch 'upstream/master' into replace-floats-with-integers

jreback commented:
yep looks ok (rebase if u can though)

simonjayhawkins commented:
> yep looks ok (rebase if u can though)

merged master an hour ago, no commits to master since then

simonjayhawkins commented:
Thanks @hasan-yaman
